Tom: fix multiple compilation errors on macOS with CUDA enabled

TomHeaven opened the pull request:
I've been maintaining a repo named [tensorflow-osx-build](https://github.com/TomHeaven/tensorflow-osx-build) publishing Tensorflow wheel packages for macOS with CUDA support for more than one year. I know there are multiple problems preventing Tensorflow from successful compilation on macOS. Therefore, I create this pull request to help solve those problems.

Now I will try to explain those problems and changes one-by-one.

## Problematic File 1: third-party/gpus/cuda_configure.bzl
+ In function `make_copy_dir_rule`, a command line parameter '-rLF' is not available on macOS and will cause an error. I replaced it with '-r -f', which is working on both Linux and macOS.
+ Function `_check_cuda_libs` is not working on macOS due to different return information of 'objdump' command. I provide a separate and working version function named _check_cuda_libs_macOS.
+ In function `_find_libs`, stub_dir should be "" for macOS.
The changes should not break anything.

## Problematic File 2: tensorflow/core/util/gpu_device_functions.h:142.
+ Function __nvvm_read_ptx_sreg_laneid() is undeclared on macOS. I removed it to solve the problem but I do not know if the change breaks anything.

## Problematic File 3: tensorflow/core/kernels/tridiagonal_solve_op_gpu.cu.cc
+ Function `SolveForSizeOneOrTwoKernel` should be a device function rather than a global function. The change should not break anything.

## Problematic File(s) 4: tensorflow/core/kernels/conv_grad_filter_ops.c, conv_grad_input_ops.cc, and conv_ops.cc.
+ Those files share the same compilation error related to the following statement
```cpp
constexpr auto kComputeInNHWC =
```
~~The compiler on macOS thinks the return type should not be constant. Thus, I removed 'constexpr' to solve the problem. The changes should not break anything.~~
Here 'constexpr' should be replaced with 'const'. The former must be determined at compilation time on macOS while the later can be determined at run time.

## Problematic File 5: tensorflow/core/platform/tstring.h.
+ Declaration  `friend  std::hash<tstring>;` raise a strange error
```
./tensorflow/core/platform/tstring.h:228:15: error: no template named 'hash'; did you mean 'std::hash'?
friend struct hash< tstring> ; 
              ^~~~
              std::hash
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:400:29: note: 'std::hash' declared here
template< class _Tp> struct hash; 
                            ^
1 error generated.
ERROR: /Volumes/Data/libraries/tensorflow/tensorflow/core/kernels/BUILD:384:1: output 'tensorflow/core/kernels/_objs/fill_functor_gpu/fill_functor.cu.pic.o' was not created
```
For now I removed it to avoid the problem but  I do not know if the change breaks anything.

## Problematic File(s) 6: tensorflow/core/kernels/data/experimental/snapshot_util.cc, snapshot_util.h
+ Constant variables 'kSnappyReaderInputBufferSizeBytes' and kSnappyReaderOutputBufferSizeBytes' defined in header class will be compiled as an indirect symbol and cause an import error on macOS, see #39262. I moved those declarations to the corresponding cc file and solved the problem. 

### Similar problems found in tensorflow/core/kernels/data/experimental/data_service_dataset_op.h and  data_service_dataset_op.cc
+ I fixed them by moving static constants within class from .h file to .cc file.
+ I suggest avoid defining static constants within class to avoid issues similar to  #39262. 

## Problematic File 7: tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:198:3
+ The error message is: error: use of undeclared identifier 'CHECK_ERR'. I removed the identifier. However, create a usable version maybe a better solution.

## Problematic File 8: tensorflow/core/util/gpu_kernel_helper.h:61
+ In macro GPU_DYNAMIC_SHARED_MEM_DECL, Key word __align__ is not available on macOS. I provide a separate version for macOS. The change should not break anything.

## An external problematic file: bazel-tensorflow/external/com_google_absl/absl/container/internal/compressed_tuple.h:168-178
+ Those two functions cause template related errors on macOS. I cannot fix them and removed them to avoid the problem. However, the sources seems to be external and I do not know where to submit my changes.

The changes are as follows (not included in this pull request):
```cpp
/*template <int I>
  ElemT<I>&& get() && {
    return std::move(*this).internal_compressed_tuple::template Storage<CompressedTuple, I>::get();
  }
  template <int I>
  constexpr const ElemT<I>&& get() const&& {
    return absl::move(*this).internal_compressed_tuple::template Storage<CompressedTuple, I>::get();
  }*/
```
Update: The problem has been fixed in master of [abseil-cpp](https://github.com/abseil/abseil-cpp). Therefore, update dependency of absl may solve the problem. More info: https://github.com/abseil/abseil-cpp/issues/392, https://github.com/tensorflow/tensorflow/issues/33489

TomHeaven committed:
Tom: fix multiple compilation errors on macOS with CUDA

TomHeaven committed:
fix multiple compilation errors on macOS

TomHeaven commented:
@googlebot I signed it!

TomHeaven committed:
fix snapshot_util.cc

TomHeaven committed:
fix gpu_kernel_helper.h

TomHeaven committed:
fix gpu_kernel_helper.h

TomHeaven committed:
use macro to detect macOS and minimize changes

TomHeaven committed:
use macro to detect macOS and minimize changes

TomHeaven committed:
fix symbol not found error in flat namespace about data_service_dataset_op

ezhulenev commented:
Thanks for the PR!

Can you pull `Problematic File(s) 4:` into a separate PR and assign it on me, no need to keep comments, just remove problematic `constexprs`. Will `static auto` compile?

Regarding `constexpr` in headers, I think it's better to follow example in: https://github.com/tensorflow/tensorflow/blob/913b88bd0d8a32115aca85aa0c1c313bd933e5ab/tensorflow/core/common_runtime/inline_function_utils.h#L234-L239

Every cc file that needs those constants redeclare them with `LowerFunctionalOpsConstants::...`. There are few example in the code.

Can you also move files under `tensorflow/core/kernels/data` into a separate PR and assign it to @jsimsa + cc me.

For cuda build rules @gunan is the owner. Also suggest to create a separate PR.

ABSL changes must go into the https://github.com/abseil/abseil-cpp.

TomHeaven commented:
Thanks for the reply. I can separate those changes (Problematic File 1, 4, 6) into different PRs.  However, how about the other changes (Problematic File 2, 3, 5, 7, 8) that you did not mention?

In fact, I need to apply all those changes to successfully compile TF on macOS. Separate them into different branches/PRs will make it inconvenient for me to test them. If it is possible,  I'd like corresponding code reviewers to look at them first. So can we invite those code reviewers to review to the PR?

TomHeaven committed:
remove comments for Problematic File(s) 4

ezhulenev commented:
@chsigg can you advice what to do with CUDA/GPU related changes here?

chsigg commented on tensorflow/core/kernels/data/experimental/snapshot_util.h:
```diff
@@ -103,11 +103,11 @@ class Reader {
   // buffer.
-  static constexpr const int64 kSnappyReaderInputBufferSizeBytes =
-      1 << 30;  // 1 GiB
+  //static constexpr const int64 kSnappyReaderInputBufferSizeBytes =
```
C++14 requires this to be defined somewhere. C++17 does not, but we still need to support C+14.

TomHeaven replied:
I moved the definition to the corresponding .cc file to avoid an import error.

chsigg commented on tensorflow/core/kernels/data/experimental/data_service_dataset_op.h:
```diff
@@ -46,7 +46,9 @@ class IterationCounter : public ResourceBase {
 class DataServiceDatasetOp : public DatasetOpKernel {
  public:
-  static constexpr const char* const kDatasetType = "DataService";
+//Tom: Avoid declare static constant variables in class, which will result in import error (Symbol not found in flat namespace) in macOS. 
```
This looks like a good change if the constants are only used in the .cc file. No need to keep the old code as a comment.

TomHeaven replied:
I'll remove the comments.

chsigg commented on tensorflow/core/kernels/conv_grad_filter_ops.cc:
```diff
@@ -838,11 +838,11 @@ void LaunchConv2DBackpropFilterOp<Eigen::GpuDevice, T>::operator()(
-
-  constexpr auto kComputeInNHWC =
+  
+  auto kComputeInNHWC =
```
Is there any way we can keep the constexpr here?

TomHeaven replied:
I don't know. The AppleClang compiler thinks returned values should not be constant.

TomHeaven replied:
I figured it out. Simply replace`constexpr` with `const` will solve the problem. According to C++11 docs, `constexpr` must be determined at compilation time, and `const` can be determined at run time.

chsigg commented on tensorflow/core/kernels/data/experimental/data_service_dataset_op.cc:
```diff
@@ -63,6 +49,22 @@ const int64 kRetryTimeoutMicros = 1000LL * 1000 * 60 * 60;  // 60 minutes.
 
+//Tom: Avoid declare static const variables in class, which will result in import error (Symbol not found in flat namespace) in macOS. 
+// Define them in CC file instead.
+  static constexpr const char* const kDatasetType = "DataService";
```
No need to declare 'static' in an anonymous namespace.

TomHeaven replied:
I removed 'static' and unnecessary comments.

chsigg commented on tensorflow/core/platform/tstring.h:
```diff
@@ -225,7 +225,23 @@ class tstring {
   friend tstring operator+(const tstring& a, const tstring& b);
   friend std::ostream& operator<<(std::ostream& o, const tstring& str);
-  friend std::hash<tstring>;
+  #if !defined(__APPLE__)
```
This looks pretty random, without understanding the underlying issue. Could you please investigate what the actual problem is?

TomHeaven replied:
I tried to define a simplest class named `tstring` with `friend std::hash<tstring>` and it can be compiled! So it's confusing why the declaration here raises the error. I'm not sure if I can find the cause.

gunan replied:
@gharibian to take a look at this. 

From what I can tell, this is your error:
error: no template named 'hash'; did you mean 'std::hash'?
so I do not think there is an issue here.

gharibian replied:
The removal may break things, as this hash is used for ordering of tstrings in stl containers.

The error messages you posted below seem to report the compilation error of the *modification( from std::hash -> struct hash. What was the original error message with this file *unaltered*?

TomHeaven replied:
@gharibian The original error message is simply without the keyword `struct` (other information is the same).  When I first encountered the error, I searched suggestions on Stackoverflow and tried adding the keyword `struct` but without luck. I'll compile the code on a different mac to see if the problem persists.

gunan replied:
For us, this seems to be working on macos.
And given that its removal will break things, we cannot accept the change in this file.

TomHeaven replied:
I tried building TF without CUDA using Xcode 11.4 and it's working.  It suggest the error may be a compiler bug of Xcode 9.4.1. Next I plan to update my compiler environment to Xcode 10.2 with CUDA 10.1. If that works, I'll revert this change.

TomHeaven replied:
I think this is a compiler bug since I didn't meet the error using CUDA10.1 with XCode 10.1. However, a new bug https://github.com/tensorflow/tensorflow/issues/40260 prevented me from successful building TF with CUDA10.1. It means I have to use CUDA 10.0 with XCode 9.4.1 (CUDA10.1 does not support newer XCode) and this compiler bug needs the workaround here until https://github.com/tensorflow/tensorflow/issues/40260 is solved.

I understand both Apple and Nvidia started to abandon macOS as a CUDA compatible platform, and there are not many users insisting on using macOS to do deep learning. I still hope to offer a chance for the 'not many users' to use TF on macOS with CUDA. So I think we should make it work first before making it perfect.

TomHeaven replied:
#40260 has been solved. Now I can revert this change.

chsigg commented on tensorflow/core/util/gpu_device_functions.h:
```diff
@@ -192,11 +192,14 @@ __device__ const unsigned kGpuWarpAll = 0xffffffff;
-#if __clang__
-  return __nvvm_read_ptx_sreg_laneid();
-#else   // __clang__
+//#if __clang__
```
That's surprising. What clang version are you using?

TomHeaven replied:
I'm using XCode 9.4.1. Here is the detail:
```
iMac18:data tomheaven$ clang++ --version
Apple LLVM version 9.1.0 (clang-902.0.39.2)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

TomHeaven replied:
Any suggestions?

TomHeaven replied:
I re-write this part and provide a separate version for macOS. At least the change does break things.

chsigg commented on tensorflow/core/util/gpu_kernel_helper.h:
```diff
@@ -56,8 +56,14 @@ using gpuError_t = hipError_t;
 
-#define GPU_DYNAMIC_SHARED_MEM_DECL(ALIGN, TYPE, NAME) \
-  extern __shared__ __align__(ALIGN) TYPE NAME[]
+#if defined(__APPLE__)
```
Simpy ignoring alignment is not going to work. There must be a way to align. Maybe with __attribute__?

TomHeaven replied:
I just tested that
```cpp
#define GPU_DYNAMIC_SHARED_MEM_DECL(ALIGN, TYPE, NAME) \
  extern __shared__  __attribute__ ((aligned(ALIGN))) TYPE NAME[]
```
is OK on macOS. I'll changed the code accordingly.

TomHeaven replied:
Unfortunately, the above declaration with alignment is only semantically good. It will lead to https://github.com/tensorflow/tensorflow/issues/14174 and trigger multiple compilation errors such as
```
tensorflow/core/kernels/concat_lib_gpu_impl.cu.cc(73): error: specified alignment (4) is different from alignment (2) specified on a previous declaration
          detected during:
            instantiation of "void tensorflow::concat_variable_kernel<T,IntType,useSmem>(tensorflow::GpuDeviceArrayStruct<const T *, 8>, tensorflow::GpuDeviceArrayStruct<IntType, 8>, IntType, IntType, T *) [with T=float, IntType=tensorflow::int32, useSmem=true]" 
(161): here
            instantiation of "void tensorflow::ConcatGPUImpl(const Eigen::GpuDevice &, const tensorflow::GpuDeviceArrayStruct<const T *, 8> &, const tensorflow::GpuDeviceArrayStruct<IntType, 8> &, __nv_bool, int, tensorflow::TTypes<T, 2, Eigen::DenseIndex>::Matrix *) [with T=float, IntType=tensorflow::int32]" 
(224): here
```

chsigg commented on tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:
```diff
@@ -195,7 +195,7 @@ static std::string GetBinaryDir(bool strip_exe) {
   char unresolved_path[buffer_size];
   _NSGetExecutablePath(unresolved_path, &buffer_size);
-  CHECK_ERR(realpath(unresolved_path, exe_path) ? 1 : -1);
+  // CHECK_ERR(realpath(unresolved_path, exe_path) ? 1 : -1);  //Tom: to sove "use of undeclared identifier 'CHECK_ERR'" on macOS
```
Can you find out why CHECK_ERR isn't available here on macOS?

TomHeaven replied:
Where is the macro CHECK_ERR supposed to be defined? I searched the whole TF project but cannot find it anywhere else. 

An interesting thing is that the part of code is to be compiled for macOS only. I wonder if is it really tested on macOS or not :)

chsigg reviewed (changes requested):
Hi Tom. Thanks for your pull request. In the current state, it seems too brittle and not quite ready to be merged. I've added some comments where more work is needed. Thanks!

gbaned commented:
@TomHeaven Can you please check @chsigg's comments and keep us posted. Thanks!

TomHeaven commented:
The comments of @chsigg helped me solved some problems. However, the changes in
+ tensorflow/core/platform/tstring.h `A compiler BUG of AppleClang?`
+ tensorflow/core/util/gpu_device_functions.h `What is CHECK_ERR for macOS?`
+ tensorflow/stream_executor/cuda/cuda_gpu_executor.cc `A macro is not defined in AppleClang`

still need help.

TomHeaven committed:
changes made according to chsigg's comments

TomHeaven committed:
revert changes of alignment

TomHeaven committed:
Merge branch 'master' into tom_macos_with_cuda

TomHeaven committed:
fix conflict with latest cuda_configure.bzl

TomHeaven committed:
fix undefined compiler parameter -fno-canonical-system-headers of cuda_configure.bzl

TomHeaven committed:
libtool of macOS does not recognize parameter 'rcsD', use '-o' instead

TomHeaven committed:
v2.2.0_mac1

TomHeaven committed:
minimize changes in tstring.h

TomHeaven committed:
revert an unnecessary changei n concat_lib_gpu_impl.cu.cc

TomHeaven committed:
Merge branch 'master' into tom_macos_with_cuda

tensorflow-jenkins committed:
Insert release notes place-fill

rchao committed:
Skip data loading error in multi_worker_tutorial_test (the test does not aim to cover this).

PiperOrigin-RevId: 317797271
Change-Id: I8336d7ffeda0836beef0a2d04e633614a44e7fa4

tensorflow-jenkins committed:
Update version numbers to 2.3.0-rc0

rxsang committed:
Support dynamic outputs for XLA on demand ops.

PiperOrigin-RevId: 317902879
Change-Id: I6b6dfa54855d5996ac15d4b5c48a5db5dc230025

goldiegadde committed:
Update setup.py

tensorflower-gardener committed:
Merge pull request #40705 from Intel-tensorflow:chuanqiw/upgrade_sqlite

PiperOrigin-RevId: 317934381
Change-Id: I95cdf789f7f5a89d75d45a1b6d67f1ad993cafab

goldiegadde committed:
Merge pull request #40734 from tensorflow/mm-cherry-pick-sqlite-fix-r2.3

Cherry-pick sqlite version bump

goldiegadde committed:
Merge pull request #40737 from geetachavan1/cherrypicks_45UX9

[CherryPick 2.3] Skip data loading error in multi_worker_tutorial_test (the test does not aim to cover this).

tensorflower-gardener committed:
Add the "--define=no_tensorflow_py_deps=true" flag for the windows cpu release
builds.

PiperOrigin-RevId: 317968971
Change-Id: I7d4db21474d85620928f3a5ffb1e4cfebaa2be9f

tanzhenyu committed:
change the size of input to remedy OOM issue.

PiperOrigin-RevId: 317995769
Change-Id: I1358449e989a41c5621e6a4d56e603387be0490d

mihaimaruseac committed:
Merge pull request #40739 from tensorflow/ggadde-cp1-2-3

[CherryPick r2.3] Add the "--define=no_tensorflow_py_deps=true" flag for the windows cp…

angerson committed:
Pin numpy below 1.19.0 in Dockerfiles

See #40688.

PiperOrigin-RevId: 318122157
Change-Id: Ief46c5610f3aaf0cdd7d43ce1a10d6d87e8e8e01

fchollet committed:
Fix a critical breakage in `training` argument default value in inference for
layers with a default of `training=True` called in e.g. a Sequential container.

PiperOrigin-RevId: 318145694
Change-Id: I1af5286824e3a45e1a7d1b8a4fadd7ec223895dc

Commit:
Allow tf.distribute.TPUStrategy to be used with TPUEmbedding API and ensure that LossScaleOptimizer properly rejects it.

PiperOrigin-RevId: 318186211
Change-Id: Id3b9cb8288e5d28ddbaec97d5b35627ab35bc08d

tensorflower-gardener committed:
Fix the api docstrings for `on_*_batch_begin` methods in `callbacks.Callback`
class.

PiperOrigin-RevId: 318203984
Change-Id: I85e2aa5d4c498b67f3e9130fa45a38173f84b35b

goldiegadde committed:
Merge pull request #40803 from geetachavan1/cherrypicks_7XBWE

[CherryPick 2.3] change the size of input to remedy OOM issue.

angerson committed:
Fix docker for numpy again

I forgot about the trouble using '<' in pip dependencies on the CLI.
This time I verified the build works.

PiperOrigin-RevId: 318332763
Change-Id: I347aee8121464232222e72d89409c600159ee80c

goldiegadde committed:
Merge pull request #40807 from geetachavan1/cherrypicks_66T6S

[CherryPick r2.3] Fix a critical breakage in `training` argument default value in inference for layers with a default of `training=True` called in e.g. a Sequential container.

goldiegadde committed:
Merge pull request #40808 from geetachavan1/cherrypicks_X2P52

[CherryPick:r2.3] Allow tf.distribute.TPUStrategy to be used with TPUEmbedding API and ensure that LossScaleOptimizer properly rejects it.

goldiegadde committed:
Merge pull request #40809 from geetachavan1/cherrypicks_NLYH0

[CherryPick:r2.3] Fix the api docstrings for `on_*_batch_begin` methods in `callbacks.Callback` class.

goldiegadde committed:
Merge pull request #40813 from geetachavan1/cherrypicks_6WCZX

[CherryPick:r2.3] Fix docker for numpy

goldiegadde committed:
Update RELEASE.md

VoVAllen committed:
Cherry pick dlpack fix into r2.3

k-w-w committed:
Add SaveableObjects to SavedModel.

When objects are loaded from the SavedModel, they don't retain their `_gather_saveables_for_checkpoint` functions, which can result in values not being loaded from the checkpoint.

This CL adds a field in the SavedModel proto that stores a save and restore function for each SaveableObject in each node. When loading into Python, the SaveableObjects are restored using the functions.

PiperOrigin-RevId: 318549786
Change-Id: I688c72d7658e1bca98abf373a13a0e15a7fb83e2

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

TomHeaven committed:
minimize change to tstring.h

VoVAllen committed:
Update tfe_wrapper.cc

VoVAllen committed:
cp

VoVAllen committed:
Merge branch 'fix_dlpack_r2.3' of https://github.com/vovallen/tensorflow into fix_dlpack_r2.3

angerson committed:
Added point about tf.sysconfig.get_build_info()

Commit:
Shift forward-compat window for NMS padded to fix TFLite conversion in 2.3

PiperOrigin-RevId: 319856402
Change-Id: I60eeb4e411609511046cc4676c2f3ad630b3769d

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Merge pull request #41002 from geetachavan1/cherrypicks_JLPTC

[CherryPick:r2.3] Support dynamic outputs for XLA on demand ops.

alextp committed:
Merge pull request #41017 from VoVAllen/fix_dlpack_r2.3

Cherry-pick dlpack fix #40843 into r2.3

goldiegadde committed:
Merge pull request #41006 from k-w-w/cherrypicks_UMQWV

Add SaveableObjects to SavedModel.

tensorflow-jenkins committed:
Update version numbers to 2.3.0-rc1

Commit:
Updated estimator version

mihaimaruseac committed:
Merge pull request #41172 from tensorflow/update_v

Updated estimator version

smit-hinsu committed:
Disable MLIR bridge for NMS image ops test

MLIR bridge doesn't support tf.NonMaxSuppressionV4 legalization that is
conditionally generated by non_max_suppression_padded function.

PiperOrigin-RevId: 320197235
Change-Id: If7242133254680b366771ced50de074ed6180563

kenfranko committed:
Correctly set the experimental_io_device when restoring variable from a checkpoint.

PiperOrigin-RevId: 320222381
Change-Id: I30187c7777ab8056e48004ef5e4ae747edc32227

Commit:
Merge pull request #41140 from srjoglekar246/cherrypicks_5EM5L

r2.3 cherry-pick request: Shift padded NMS compat window forward to fix TFLite conversion

fchollet committed:
Fix critical bug with `add_loss` TFOpLayer graph construction
that caused incorrect loss values and backprop issues

PiperOrigin-RevId: 320257330
Change-Id: I0a030bc7632735b152454657fd15e41539b4e4bd

goldiegadde committed:
Update RELEASE.md

caisq committed:
Add mentioned Debugger V2 to r2.3 release notes

caisq committed:
Grammar tweaks in the Debugger V2 bullet point

gbaned commented:
@chsigg  Can you please review this PR ? Thanks!

gunan commented:
There are a few issues here.
@gharibian wdyt about the change in tstring?
If it is a non-starter, we may want to close the PR.

Commit:
Explicitly raise a (clearer) error message when models end up in invalid states due to interleaving graph and eager.

In rare cases code may have run w/o crashing when in these invalid states, but it's safer to error with an explanation rather than risk silent failures/fragile behavior.

PiperOrigin-RevId: 321192744
Change-Id: I9e97ac3b7cea27c9b389e5202de9f1c09a4aa2b8

goldiegadde committed:
Update RELEASE.md

goldiegadde committed:
Update RELEASE.md

mihaimaruseac committed:
Fix typo in estimator min bound

Should be `2.3.0rc0`, not `2.3.0-rc0`

mihaimaruseac committed:
Merge pull request #41435 from tensorflow/Fix-Estimator-min-bound

Fix typo in estimator min bound

rohan100jain committed:
Going back to forcing embedding layer variables on the CPU even within a tf.function as this is breaking some user code.

PiperOrigin-RevId: 321607029
Change-Id: Id159867f51b26e6604a1186d9ce526658ddd1e19

aaudiber committed:
Update "master" to "dispatch"/"dispatcher" in tf.data service terminology.

Dispatcher is more descriptive and follows the guidance in https://developers.google.com/style/word-list#master

PiperOrigin-RevId: 321613785
Change-Id: Iaa576d35f0581e21278101f8b31201ba737a6865

Commit:
Raise an error when some but not all values passed to the first layer call arg are symbolic. This setting can cause functional models to be constructed incorrectly.

Support for this will be added when we enable the KerasTensors refactoring.

Addreses GitHub Issue #40638

PiperOrigin-RevId: 321639068
Change-Id: Iebf0e1198018fe44b1f60673bd991a9262ecef7d

Commit:
Relax the error about functional api construction w/ a mix of symbolic and non-symbolic tensors for built-in layers (such as layers.add and layers.multiply where using constants is a common user pattern)

PiperOrigin-RevId: 321698209
Change-Id: Ief13e59aec91b787361a7760318ecd47870d938f

goldiegadde committed:
Merge pull request #41426 from geetachavan1/cherrypicks_MI0K1

[Cherrypick r2.3] Fix critical bug with `add_loss` TFOpLayer graph construction that caused incorrect loss values and backprop issues

goldiegadde committed:
Merge pull request #41385 from tomerk/cherrypicks_ORV3K

[Cherrypick:r2.3] Explicitly raise a (clearer) error message when models end up in inva…

tensorflow-jenkins committed:
Update version numbers to 2.3.0-rc2

mihaimaruseac committed:
Merge pull request #41464 from aaudiber/cherrypicks_DM5NO

Update "master" to "dispatch"/"dispatcher" in tf.data service terminology

mihaimaruseac committed:
Merge pull request #41357 from kenfranko/cherrypicks_4ECL0

2.3-rc2 cherry-pick request: Correctly set the experimental_io_device when restoring variable from a checkpoint.

goldiegadde committed:
Merge pull request #41483 from tomerk/cherrypicks_DC0YA

Raise an error when some but not all values passed to the first layer…

goldiegadde committed:
Merge pull request #41498 from tensorflow-jenkins/version-numbers-2.3.0rc2-11909

Update version numbers for TensorFlow 2.3.0-rc2

gbaned commented:
@gharibian, @gunan Any update on this PR? Please. Thanks!

gharibian commented:
> @gharibian, @gunan Any update on this PR? Please. Thanks!

@gbaned The root cause for the compilation issues w.r.t. tstring needs to be isolated; placing blanket preprocessor guards is not an acceptable solution without knowing the cause--especially since std::hash<tstring> is used by stl containers for ordering.  std::hash is a part of the standard, so the reason for the breakage needs to be isolated.

Please refer back to my last post for suggestions on moving forward.

angerson committed:
Reword tf.sysconfig.get_build_info notice

An internal bug report revealed that "support" could be mistaken for "official support," which is not intended.

tensorflower-gardener committed:
Update tensorboard dependency to 2.3.x

TensorBoard release: https://pypi.org/project/tensorboard/2.3.0/

PiperOrigin-RevId: 322474571
Change-Id: I7cb6bcbb101cb9b10d04b832279e62ea9066abca

TomHeaven committed:
revert changes to tstring.h

TomHeaven commented:
> > @gharibian, @gunan Any update on this PR? Please. Thanks!
> 
> @gbaned The root cause for the compilation issues w.r.t. tstring needs to be isolated; placing blanket preprocessor guards is not an acceptable solution without knowing the cause--especially since std::hash is used by stl containers for ordering. std::hash is a part of the standard, so the reason for the breakage needs to be isolated.
> 
> Please refer back to my last post for suggestions on moving forward.

I think the root cause of the compilation error is a compiler bug in Apple's XCode 9.4.1 compiler. The problem does not exist  when I upgrade to XCode 10.1 (which requires at least CUDA 10.1 on macOS). I revert the change to meet your requirement. However, that code is not able to be compiled with CUDA 10.0 (which is requires XCode 9.4.1 on macOS).

psybuzz committed:
Merge pull request #41633 from psybuzz/cherrypicks_WRXIO

This cherrypick is required to depend on a version of TensorBoard that
is compatible with TensorFlow 2.3.x, such that `tensorboard` starts
successfully.

Note: starting from TF/TB 2.3, the tensorboard dependency is more
relaxed to allow TensorBoard to be released at a more frequent cadence.
The major version is still synced, while the minor version is not.

PiperOrigin-RevId: 322474571

goldiegadde committed:
Merge pull request #41636 from geetachavan1/cherrypicks_F4YCK

[CherryPick 2.3] Going back to forcing embedding layer variables on the CPU even within a tf.function as this is breaking some user code.

goldiegadde committed:
Update RELEASE.md

Commit:
Updating version for final release

Commit:
Updating version for final release

mihaimaruseac committed:
Merge pull request #41644 from tensorflow/update_v

Updating estimator version after estimator final release

tensorflow-jenkins committed:
Update version numbers to 2.3.0

mihaimaruseac committed:
Merge pull request #40701 from tensorflow-jenkins/relnotes-2.3.0rc0-13382

Update release notes for TensorFlow 2.3.0

goldiegadde committed:
Merge pull request #41670 from tensorflow-jenkins/version-numbers-2.3.0-24378

Update version numbers for TensorFlow 2.3.0

TomHeaven committed:
Merge commit 'b36436b087bd8e8701ef51718179037cccdfc26e' into tom_macos_with_cuda

gbaned commented:
@TomHeaven Can you please sign CLA and resolve conflicts?

TomHeaven commented:
@googlebot I signed it!

TomHeaven commented:
@googlebot I consent.

TomHeaven commented:
> @googlebot I consent.

@googlebot I consent.

TomHeaven commented:
@googlebot I consent.

gbaned commented:
@TomHeaven, Any update on this PR? Please. Thanks!

gbaned commented:
@TomHeaven Any update on this PR? and please resolve conflicts. Thanks!

jaingaurav commented:
Thank you for these fixes. However, I don't think we can review this PR in its current state. It currently tries to solve too many different issues which requires different expertise and reviews across the stack. Even if we were able to reach consensus it would be very challenging to get these to pass all internal tests as each of these fixes might have their own set of subtle failures. Please break up the CL and include the necessary tests for the respective fixes.
